### PR TITLE
Port the MainWindow ui to a .ui file

### LINKED ISSUE #69: Widen instance management buttons
### Role

I use the launcher

### Suggestion

I want the buttons on the right to have wider "hitboxes"

### Benefit

so I can more easily click the "Edit" button (that was recently shortened in size from "Edit instance") with more ease.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Снимок
Hovering over "Edit" (the most important button in the list) shows how actually small the button is. I'd like it to span for the entirety of the sidebar's width no matter how short the button text is.

### LINKED ISSUE #473: Fix tooltip
https://github.com/PrismLauncher/PrismLauncher/blob/347ae0a9ad8d5d933ef76dedab177995029e5f0d/launcher/ui/MainWindow.cpp#L613

### LINKED ISSUE #594: Extend the sidebar buttons
As of 6.0, the buttons for the selected instance are only as long as their text - it would be a welcome change to extend them all the way to the right.

### PR DESCRIPTION
**Pull request by leia-uwu**
Some stuff still needs to be done in the c++ side because qt designer is dumb
(The instance toolbar icon and name buttons, and some shortcuts are still manually added in MainWindow.cpp)

Looks almost identical, with some minor tweaks:
- The instance toolbar has been changed to a WideBar, allowing for customization of actions.
- The instance toolbar buttons are now fullwidth.
- The close window action has been moved to the end of the file menu and now has an icon.
- The help menu has some layout changes.

This also fixes some stuff:
- Menus not having tooltips.
- The top toolbar not connecting to the title bar in kde.
- The instance toolbar separators looking weird after you move the toolbar.

Closes #594, closes #69, closes #473

### CODE REVIEW

File: launcher/ui/MainWindow.cpp [hunk 1]
```diff
@@ -1375,14 +698,10 @@ void MainWindow::repopulateAccountsMenu()
 
     if (accounts->count() <= 0)
     {
-        ui->all_actions.removeAll(&ui->actionNoAccountsAdded);
-        ui->actionNoAccountsAdded = TranslatedAction(this);
-        ui->actionNoAccountsAdded->setObjectName(QStringLiteral("actionNoAccountsAdded"));
-        ui->actionNoAccountsAdded.setTextId(QT_TRANSLATE_NOOP("MainWindow", "No accounts added!"));
+        ui->actionNoAccountsAdded->setText( "No accounts added!");
```
Comment (flowln) [inline launcher/ui/MainWindow.cpp]:
```suggestion
        ui->actionNoAccountsAdded->setText(tr("No accounts added!"));
```

File: launcher/ui/MainWindow.cpp [hunk 2]
```diff
@@ -1414,18 +733,17 @@ void MainWindow::repopulateAccountsMenu()
...
     accountMenu->addSeparator();
-    ui->profileMenu->addSeparator();
+    ui->accountsMenu->addSeparator();
 
-    ui->all_actions.removeAll(&ui->actionNoDefaultAccount);
-    ui->actionNoDefaultAccount = TranslatedAction(this);
+    ui->actionNoDefaultAccount = new QAction(this);
     ui->actionNoDefaultAccount->setObjectName(QStringLiteral("actionNoDefaultAccount"));
-    ui->actionNoDefaultAccount.setTextId(QT_TRANSLATE_NOOP("MainWindow", "No Default Account"));
+    ui->actionNoDefaultAccount->setText("No Default Account");
```
Comment (flowln) [inline launcher/ui/MainWindow.cpp]:
```suggestion
    ui->actionNoDefaultAccount->setText(tr("No Default Account"));
```

File: launcher/ui/MainWindow.cpp [hunk 3]
```diff
@@ -139,785 +140,100 @@ QString profileInUseFilter(const QString & profile, bool used)
...
-        if (MainWindow->m_statusCenter)
-        {
-            MainWindow->updateStatusCenter();
-        }
-    } // retranslateUi
-};
+        // disabled until we have an instance selected
+        ui->instanceToolBar->setEnabled(false);
+        ui->actionKillInstance->setEnabled(false);
+        ui->actionLaunchInstance->setEnabled(false);
```
Comment (flowln) [inline launcher/ui/MainWindow.cpp]: those two don't seem to have any impact, removing them still leaves the actions disabled when no instance is selected :p

File: launcher/ui/MainWindow.cpp [hunk 4]
```diff
@@ -929,12 +244,13 @@ MainWindow::MainWindow(QWidget *parent) : QMainWindow(parent), ui(new MainWindow
...
         m_newsChecker.reset(new NewsChecker(APPLICATION->network(), BuildConfig.NEWS_RSS_URL));
-        newsLabel = new QToolButton();
-        newsLabel->setIcon(APPLICATION->getThemedIcon("news"));
-        newsLabel->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
-        newsLabel->setToolButtonStyle(Qt::ToolButtonTextBesideIcon);
-        newsLabel->setFocusPolicy(Qt::NoFocus);
-        ui->newsToolBar->insertWidget(ui->actionMoreNews, newsLabel);
+        newsLabel = dynamic_cast<QToolButton*>(ui->newsToolBar->widgetForAction(ui->actionNewsLabel));
+
+        //add a spacer before the more news button
```
Comment (flowln) [inline launcher/ui/MainWindow.cpp]: i think the previous behavior was a bit better, you had a lot more room to click the news entry if you wanted, space that otherwise is just empty.

**Review by flowln — COMMENTED** [review]
this looks pretty good overall!

there's only a couple minor things that I think could be improved, outside the comments i've made:
1. While you did fix the alt shortcuts on the WideBar, they still show up in the Right-click menu, even though you can't even use those without closing the menu :p
image

2. I'm not sure about having icons in that menu. In the instance WideBars, this was not an issues because the actions don't have icons, but here they do, and it becomes ambiguous whether clicking that option will toggle its visibility, or activate the action.

3. Maybe the old right-click menu actions, now only accessible through the top bar and the news bar, should be merged with this new menu somehow. It is quite weird to not be able to hide the Instance toolbar from interacting with that toolbar, but be able to by interacting with other toolbars. Perhaps they could be separated into two different sections, and the one to toggle the actions visibility have a name that would clear up the ambiguity I talked about in item 2, so the icons could be left there :)

**Comment by leia-uwu** [comment]
> 1. even though you can't even use those without closing the menu :p

uh, for some reason the menu only closes when you press alt while using the fusion theme in my testing, this also happens in the instance view context menu and in the menubar :/

**Comment by leia-uwu** [comment]
:catstare:
image

**Comment by DioEgizio** [comment]
with this the status bar doesn't seem to follow my prism theme

this pr this pr

w/o this pr without this pr

**Comment by TayouVR** [comment]
hmmm... its a different shade of grey and doesn't have the margins/paddings from the theme. 
After scrolling through the code I can't spot any obvious reasons for that, maybe the various size constraints on the QStatusBar in the .ui are different from the defaults it had in code? 🤔 
image

**Comment by leia-uwu** [comment]
> with this the status bar doesn't seem to follow my prism theme

Fixed

File: launcher/ui/MainWindow.cpp (hunk 4 above)
Comment (leia-uwu) [reply launcher/ui/MainWindow.cpp]: idk how to fix that without adding the news label toolbutton manually, i tried setting the size policy but it has no effect :/

Comment (flowln) [reply launcher/ui/MainWindow.cpp]: seems like it indeed will not be possible without manually inserting the toolbutton, since it gets added via `addAction` in the generated file, which i think forces the layout to be that way. Still, I think it is still worth it to manually add it :slightly_smiling_face:

**Review by Scrumplex — APPROVED** [review]
I couldn't find any differences, besides the news toolbar as @flowln already pointed out.
LGTM and thanks!

**Review by TayouVR — APPROVED** [review]
couldn't find any issues in my testing, looking good!

File: launcher/ui/MainWindow.cpp [hunk 5]
```diff
@@ -139,785 +140,101 @@ QString profileInUseFilter(const QString & profile, bool used)
...
 
-    void setupUi(MainWindow *MainWindow)
+    // set the menu for the folders and help tool buttons
     {
-        if (MainWindow->objectName().isEmpty())
-        {
-            MainWindow->setObjectName(QStringLiteral("MainWindow"));
-        }
-        MainWindow->resize(800, 600);
-        MainWindow->setWindowIcon(APPLICATION->getThemedIcon("logo"));
```
Comment (kumquat-ir) [inline launcher/ui/MainWindow.cpp:847]: This seems to have been removed without a replacement, the window icon is missing (left is this pr, right is `aur/prismlauncher-git`)
Screenshot_2023-01-17_02-27-38

File: launcher/ui/widgets/WideBar.cpp [hunk 6]
```diff
@@ -133,7 +158,9 @@ void WideBar::insertSeparator(QAction* before)
         return;
 
     BarEntry entry;
-    entry.bar_action = QToolBar::insertSeparator(before);
+    entry.bar_action = new QAction("", this);
+    entry.bar_action->setSeparator(true);
+    insertAction(iter->bar_action, entry.bar_action);
```
Comment (flowln) [inline launcher/ui/widgets/WideBar.cpp]: why is this change needed?

File: launcher/ui/widgets/WideBar.cpp [hunk 7]
```diff
@@ -111,6 +126,16 @@ void WideBar::insertActionAfter(QAction* after, QAction* action)
...
 }
 
+void WideBar::insertWidgetBefore(QAction* before, QWidget* widget)
+{
+    auto iter = getMatching(before);
+    if (iter == m_entries.end())
+        return;
+
+    BarEntry entry;
+    entry.bar_action = insertWidget(iter->bar_action, widget);
```
Comment (flowln) [inline launcher/ui/widgets/WideBar.cpp]: what's the use of `entry` here?

File: launcher/ui/MainWindow.ui [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+  <action name="actionKillInstance">
+   <property name="icon">
+    <iconset theme="status-bad">
+     <normaloff>.</normaloff>.</iconset>
+   </property>
+   <property name="text">
+    <string>&amp;Kill</string>
+   </property>
+   <property name="toolTip">
+    <string>Kill the running instance</string>
```
Comment (flowln) [inline launcher/ui/MainWindow.ui]:
```suggestion
    <string>Kill the running instance.</string>
```

File: launcher/ui/MainWindow.ui [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+  <action name="actionMATRIX">
+   <property name="icon">
+    <iconset theme="matrix">
+     <normaloff>.</normaloff>.</iconset>
+   </property>
+   <property name="text">
+    <string>&amp;Matrix Space</string>
+   </property>
+   <property name="toolTip">
+    <string>Open %1 Matrix space</string>
```
Comment (flowln) [inline launcher/ui/MainWindow.ui]:
```suggestion
    <string>Open %1 Matrix space.</string>
```

**Review by flowln — COMMENTED** [review]
some nitpicks, but looks good overall!

**Comment by edgars-cirulis** [comment]
Kind of annoying, after toggling, the check automatically closes the menu.

File: launcher/ui/widgets/WideBar.cpp (hunk 6 above)
Comment (leia-uwu) [reply launcher/ui/widgets/WideBar.cpp]: it was adding the separator to the end of the toolbar instead of before the action

File: launcher/ui/widgets/WideBar.cpp (hunk 7 above)
Comment (leia-uwu) [reply launcher/ui/widgets/WideBar.cpp]: oops

Comment (leia-uwu) [reply launcher/ui/widgets/WideBar.cpp]: fixed

File: launcher/ui/widgets/WideBar.cpp (hunk 6 above)
Comment (flowln) [reply launcher/ui/widgets/WideBar.cpp]: that's because the old call was using `before` as the action to insert in relation to, i believe. It should use `iter->bar_action` instead, since that's the actual action in the WideBar.

Comment (leia-uwu) [reply launcher/ui/widgets/WideBar.cpp]: done

**Review by flowln — APPROVED** [review]
lgtm, thanks!
